docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,249 @@
+  the cluster and all uses of the previous version to finish.
+
+This RFC is focused on how to wait for the table descriptor change
+propagate to all nodes in the cluster. More accurately, it is focused
+on how to determine when the previous version of the descriptor is no
+longer in use. Additionally, we need mechanisms to ensure that
+multiple updates to a table descriptor are not performed concurrently
+and that write operations never commit once a table descriptor is too
+old (i.e. not one of the two most recent versions).
+
+At a high-level, we want to provide read and write leases for table
+descriptors. When a schema change is to be performed, the operation
+first needs to grab a write lease for the descriptor. The write lease
+ensures that no concurrent schema change will be performed and that
+there is only one active version of the descriptor in the cluster.
+Note that these are essentially the same constraint: there are only
+two active versions of a descriptor during a schema change.
+
+Before using a table descriptor for a DML operation (i.e. `SELECT`,
+`INSERT`, `UPDATE`, `DELETE`, etc), the operation needs to obtain a
+read lease for the descriptor. The lease will be guaranteed valid for
+a significant period of time (on the order of minutes) and the lease
+period can be extended as long as a scheme change is not
+performed. When the operation completes it will release the read
+lease.
+
+# Detailed design
+
+Table descriptors will be extended with a version number that is
+incremented on every change to the descriptor:
+
+```proto
+message TableDescriptor {
+  ...
+  optional uint32 id;
+  optional uint32 version;
+  ...
+}
+```
+
+Leases will be tied to a specific version of the descriptor. Lease
+state will be stored in a new `system.lease` table:
+
+```sql
+CREATE TABLE system.lease (
+  DescID     INT,
+  Version    INT,
+  NodeID     INT,
+  Expiration TIMESTAMP,
+  PRIMARY KEY (DescID, Version, NodeID, Expiration)
+)
+```
+
+Entries in the lease table will be added and removed as leases are
+acquired and released. A background goroutine will periodically delete
+expired leases. 
+
+Leases will be granted for a duration measured in minutes (we'll
+assume 5m for the rest of this doc, though experimentation may tune
+this number). A node will acquire a lease before using it in an
+operation and may release the lease when the last local operation
+completes that was using the lease.
+
+The leader of the range containing a table descriptor will gossip the
+most recent version of that table descriptor using the gossip key
+`"table-<descID>"` and the value will be the version number. The
+gossiping of the most recent table versions allows nodes to
+asynchronously discover when a new table version is written. But note
+that it is not necessary for correctness as the protocol for acquiring
+a lease ensures that only the most recent version of a descriptor can
+have a new lease granted on it.
```

Comment:
yes, now zoom in to

> index transitioned to write-only mode, backfill index, index transitioned to read-write mode

and you're in the situation that I'm talking about. The node is in write-only and tries to backfill, if that takes too long it's going to be forced to use the index in read-write which is going to be incorrect since the backfill hasn't completed yet.